bgpd/bgp_updgrp_adv.c

```diff
@@ -543,9 +544,14 @@ void bgp_adj_out_set_subgroup(struct bgp_dest *dest,
 	 * the route wasn't changed actually.
 	 * Do not suppress BGP UPDATES for route-refresh.
 	 */
-	if (CHECK_FLAG(bgp->flags, BGP_FLAG_SUPPRESS_DUPLICATES)
-	    && !CHECK_FLAG(subgrp->sflags, SUBGRP_STATUS_FORCE_UPDATES)
-	    && adj->attr_hash == attr_hash) {
+	if (CHECK_FLAG(bgp->flags, BGP_FLAG_SUPPRESS_DUPLICATES) &&
+	    !CHECK_FLAG(subgrp->sflags, SUBGRP_STATUS_FORCE_UPDATES)
+	    /* When the adj entry for dedup check was not there, dedup should not happen */
+	    && !no_dedup
+	    /* Is not withdraw */
+	    && !adj->is_withdraw
+	    /* Still have to compare the original attr to make sure nothing has changed */
+	    && adj->attr && attr && attrhash_cmp(adj->attr, attr)) {
```

Comment:
`adj->attr && attr` seems redundant: `attr` can't be NULL here.

Reply:
There are so many corner cases, I added the check anyways just to make sure no core dump could happen.

Reply:
IMO, this should never happen, otherwise we have more problems.